assembly/src/parsers/ast/field_ops.rs

```diff
@@ -1,12 +1,96 @@
 use super::{
-    super::parse_bit_len_param, super::parse_element_param, AssemblyError, Instruction, Node,
-    Token, Vec,
+    super::parse_element_param,
+    AssemblyError,
+    Instruction::*,
+    Node::{self, Instruction},
+    Token,
 };
-use crate::validate_operation;
 
+// INSTRUCTION PARSERS
+// ================================================================================================
+
+/// Returns `Add` instruction node if no immediate value is provided or `AddImm` instruction
+/// node otherwise.
+///
+/// # Errors
+/// Returns an error if the instruction token has invalid param or more than one param.
+pub fn parse_add(op: &Token) -> Result<Node, AssemblyError> {
+    debug_assert_eq!(op.parts()[0], "add");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Ok(Instruction(Add)),
+        2 => {
+            let imm = parse_element_param(op, 1)?;
+            Ok(Instruction(AddImm(imm)))
+        }
+        _ => Err(AssemblyError::extra_param(op)),
+    }
+}
+
+/// Returns `Sub` instruction node if no immediate value is provided or `SubImm` instruction
+/// node otherwise.
+///
+/// # Errors
+/// Returns an error if the instruction token has invalid param or more than one param.
+pub fn parse_sub(op: &Token) -> Result<Node, AssemblyError> {
+    debug_assert_eq!(op.parts()[0], "sub");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Ok(Instruction(Sub)),
+        2 => {
+            let imm = parse_element_param(op, 1)?;
+            Ok(Instruction(SubImm(imm)))
+        }
+        _ => Err(AssemblyError::extra_param(op)),
+    }
+}
+
+/// Returns `Mul` instruction node if no immediate value is provided or `MulImm` instruction
+/// node otherwise.
+///
+/// # Errors
+/// Returns an error if the instruction token has invalid param or more than one param.
+pub fn parse_mul(op: &Token) -> Result<Node, AssemblyError> {
+    debug_assert_eq!(op.parts()[0], "mul");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Ok(Instruction(Mul)),
+        2 => {
+            let imm = parse_element_param(op, 1)?;
+            Ok(Instruction(MulImm(imm)))
+        }
```

Comment:
We had an optimization to either produce `Noop` or no token at all if we multiply by special values such as `1` or `0`. Should we keep these?

Reply:
We handle these at runtime. We could do it here too but it would be a bigger refactoring, primarily because we don't have `incr` or `noop` instructions at the AST level (these are available only at MAST level).